### PR TITLE
SVG-related improvements and fixes

### LINKED ISSUE #369: Path.toPathData and Path.toSVG - X Axis is flipped
I'm using OpentypeJS to create SVGs for single glyph characters. However, when I draw the SVGs, the glyph is flipped on it's X axis ( upside down ). Even the ./bin/test-render in the repo generates flipped SVG path data.

## Expected Behavior

1) Load font
2) Get glyph path data with glyph.path.toSVG()
3) Use the path data to render proper glyph

## Current Behavior

1) Load font
2) Get glyph path data with glyph.path.toSVG()
3) Use the path data to render glyph, the glyph is rendered upside down.

## Possible Solution

Fix path commands. 

## Steps to Reproduce (for bugs)

Just run the /bin/test-render  from the repo

## Your Environment

* Version used:  0.11.0
* Font used: Every font

### LINKED ISSUE #526: Support using a plain SVG file as a glyph
## Expected Behavior

I don't know if this would be considered out of scope or not, but I'd really like if this library supported importing a raw SVG string as a glyph, which would cut out all sorts of intermediary dependencies one currently needs when generating an .odt font.

## Current Behavior

It doesn't seem possible to import raw SVG.

## Possible Solution

Maybe the Glyph class could get an additional `svg` property that internally extracts the path from the SVG?

### PR DESCRIPTION
**Pull request by Connum**
This PR supersedes draft #568

## Description

* implements loading of SVG path data into Path(), including basic modification like offset and scaling (fix #526)
* fixes SVG paths being flipped upside-down (fix #369)
* Helps the library pass more [unicode-org/text-rendering-tests](https://github.com/unicode-org/text-rendering-tests/issues) tests (see #326)
  * implements path optimizations for shorter/smaller path data *
  * improves the SVG rendering in the test-render script *

## Motivation and Context

The library failed several tests in the unicode test tool, despite technically supporting the underlying features.
Exporting SVG paths was possible, but the paths were flipped upside down due to inverted Y-axes in SVG vs font paths. Importing SVG paths was not supported.

## How Has This Been Tested?

We now pass 69 tests instead of 37 in the unicode test tool, see report: https://connum.github.io/opentypejs-reports/reports/test-connum_svg-optimize-7e4de2c2ce9e822161b61349cf2a7ec55c7fd05e.html
We didn't have any tests for path.js so I added several to test the basic and new features.
I played around with loading, exporting and modifying SVG paths in existing fonts.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
Will break code that relies on SVG data being upside down, but that behaviour was wrong in the first place

## Checklist:

- [X] I did `npm run test` and all tests passed green (including code styling checks).
- [X] I have added tests to cover my changes.
- [X] My change requires a change to the documentation.
- [X] I have updated the **README** accordingly.
- [X] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: src/path.js [hunk 1]
```diff
@@ -16,6 +16,283 @@ function Path() {
...
+    }
+    commands = [].concat.apply([], subpaths); // flatten again
+    return commands;
+}
+
+/**
+ * Returns options merged with the default options for parsing SVG data
+ * @param {object} options (optional)
+ */
+function defaultSVGParsingOptions(options) {
```
Comment (ILOVEPIE) [inline src/path.js:65]: wouldn't it be better to call this createSVGParsingOptions, because if we're overriding the defaults it's not default.

File: src/path.js [hunk 2]
```diff
@@ -16,6 +16,283 @@ function Path() {
...
+    };
+    const newOptions = Object.assign({}, defaultOptions, options);
+    return newOptions;
+}
+
+/**
+ * Returns options merged with the default options for outputting SVG data
+ * @param {object} options (optional)
+ */
+function defaultSVGOutputOptions(options) {
```
Comment (ILOVEPIE) [inline src/path.js:82]: Same here.

File: src/path.js [hunk 3]
```diff
@@ -16,6 +16,283 @@ function Path() {
...
+        });
+    }
+
+    function makeRelative(buffer) {
+        if (!this.commands.length) {
+            return buffer;
+        }
+        const lastCommand = this.commands[this.commands.length - 1];
+        for (let i = 0; i < buffer.length; i++) {
+            buffer[i] += lastCommand[i % 2 ? 'y' : 'x'];
```
Comment (ILOVEPIE) [inline src/path.js:138]: Just a small optimization, ``x % 2`` can always be rewritten as ``x & 1`` which runs faster, not that big of a deal, but processors are a lot slower at modulus (due to the division).

**Review by ILOVEPIE — CHANGES_REQUESTED** [review]
I'd like to see someone else's opinion on this, the changes I've suggested are relatively minor, this is however a breaking change PR, so we should probably lump it into 2.0.0

File: src/path.js (hunk 1 above)
Comment (Connum) [reply src/path.js:65]: True, I'll change that!

File: src/path.js (hunk 3 above)
Comment (Connum) [reply src/path.js:138]: Interesting! That's good to know. We could probably optimize performance further by using a generator function and yield, but our current build process doesn't transpile that. I'll add that as a todo comment.

**Comment by Connum** [comment]
> this is however a breaking change PR, so we should probably lump it into 2.0.0
I was hesitant whether to mark it as breaking or not. Having upside down output is obviously not what anyone would expect, on the other hand it's been like that for years, so there's probably code out there that relies on it and flips it itself.

File: src/path.js [hunk 4]
```diff
@@ -16,6 +16,285 @@ function Path() {
...
+            throw new Error('Unexpected character: ' + token + ' at offset ' + i);
+        }
+    }
+    applyCommand.apply(this);
+
+    if (options.optimize) {
+        this.commands = optimizeCommands(this.commands);
+    }
+
+    let flipY = options.flipY;
```
Comment (yne) [inline src/path.js:268]: is this variable a number or a boolean ?

File: src/path.js [hunk 5]
```diff
@@ -16,6 +16,285 @@ function Path() {
...
+
+    return this;
+};
+
+/**
+ * Generates a new Path() from an SVG path element or path notation
+ * @param  {string|SVGPathElement}
+ * @param  {object}
+ */
+Path.fromSVG = function(path, options) {
```
Comment (yne) [inline src/path.js:293]: Don't offer much gain, not sure about it usefulness

File: src/path.js (hunk 4 above)
Comment (Connum) [reply src/path.js:268]: See options documentation: "whether to flip the Y axis of the path data, because SVG and font paths use inverted Y axes. (true: calculate from bounding box, false: disable, integer/float: a base value for the calculation. default: true)"

It can be useful to calculate the actual base value from the font's ascender and descender values or the viewbox of an SVG you want to add the generated path to.

File: src/path.js (hunk 5 above)
Comment (Connum) [reply src/path.js:293]: Well, we use such shorthands in other places of the codebase too, and I liked the shorter notation without having to create an empty Path first.

File: src/path.js [hunk 6]
```diff
@@ -16,6 +16,285 @@ function Path() {
...
+function createSVGOutputOptions(options) {
+    // accept number for backwards compatibility
+    // and in that case set flipY to false
+    if (parseInt(options) === options) {
+        options = { decimalPlaces: options, flipY: false };
+    }
+    const defaultOptions = {
+        decimalPlaces: 2,
+        optimize: true,
+        flipY: true
```
Comment (yne) [inline src/path.js:91]: we shall keep it `false`, and flip it back to `true` in 2.0.0
so we could merge this PR now

Comment (Connum) [reply src/path.js:91]: Is there a plan to release a 1.x version before 2.0.0? Because we already have the breaking change regarding the platform-agnostic `font.table.name` values merged in our main branch

File: src/path.js [hunk 7]
```diff
@@ -230,17 +509,18 @@ Path.prototype.draw = function(ctx) {
...
  * See http://www.w3.org/TR/SVG/paths.html#PathData
- * @param  {number} [decimalPlaces=2] - The amount of decimal places for floating-point values
+ * @param  {object|number} [options={decimalPlaces:2, optimize:true}] - Options object (or amount of decimal places for floating-point values for backwards compatibility)
  * @return {string}
  */
-Path.prototype.toPathData = function(decimalPlaces) {
-    decimalPlaces = decimalPlaces !== undefined ? decimalPlaces : 2;
+Path.prototype.toPathData = function(options) {
+    // set/merge default options
+    options = createSVGOutputOptions(options);
```
Comment (yne) [inline src/path.js:517]: no need for a separate function, I think this wold also work
```js
Path.prototype.toPathData = function(options = 2) {
  const def = { decimalPlaces: 2, optimize: true, flipY: true};
  options = Object.assign(def, isNaN(options) ? options : {decimalPlaces:options});
}
```

Comment (Connum) [reply src/path.js:517]: My code was structured differently at first, and I wanted to be able to use the options in the glyph.js wrapper functions as well, in order to pass the base value for flipY from the glyph metrics and already have the default values there. But I didn't want to make things more complicated with this PR as they already are. We might need that in the future though, that's why I left them in a function.

File: src/path.js (hunk 4 above)
Comment (yne) [reply src/path.js:268]: I was referring to the left side `let flipY`

File: src/path.js (hunk 6 above)
Comment (yne) [reply src/path.js:91]: We need to clarify this road map (with @fdb ?)
If the next versions is going to be 2.0.0(-alpha1) then no problem for me

**Review by yne — APPROVED** [review]
Very nice addition, but we shall clarify ASAP if the next release is going to be 2.0.0, in which case some function argument could be completely migrated (number => object).

File: src/path.js (hunk 4 above)
Comment (Connum) [reply src/path.js:268]: It just takes on the value of the corresponding option. I can rewrite it if that's considered bad practice?

Comment (yne) [reply src/path.js:268]: I feel a litle bit confused when I see
```js
let flipY = options.flipY // boolean
```
then
```js
flipY = boundingBox.y1 + boundingBox.y2 // number ?!
```
but
```js
flipY !== false // boolean
```
and
```js
flipY - cmd[prop] // number again
```
I think there is 2 usecase for the same variable
the "offset" (number) version would be better in a separate variable I think.

File: src/path.js (hunk 3 above)
Comment (Connum) [reply src/path.js]: Chromium browsers seem to do some optimizations that make the difference marginal (and sometimes modulus runs even a tiny bit better than bitwise), but the story is different in Firefox. I had differences of up to ~15%:

https://jsbench.me/2zlehfyyi2/1

image

Comment (yne) [reply src/path.js]: I confirm that, on chromium-based browser get almost identical result for both `%` and `&`.

The PR is (re)open so you can merge it (no issue for me :+1: ).

Let's just not forget that `premature optimization is the root of all evil` ;)

File: src/path.js (hunk 4 above)
Comment (Connum) [reply src/path.js:268]: Makes sense that it might be confusing. Changed it with a new commit.

Comment (Connum) [reply src/path.js:268]: Woops, pushed to the wrong repo, but now it's in.

File: src/path.js [hunk 8]
```diff
@@ -16,6 +16,286 @@ function Path() {
...
+
+    if (options.optimize) {
+        this.commands = optimizeCommands(this.commands);
+    }
+
+    let flipY = options.flipY;
+    const doFlipY = !!flipY || flipY === 0;
+    if (options.flipY === true) {
+        const boundingBox = this.getBoundingBox();
+        flipY = boundingBox.y1 + boundingBox.y2;
```
Comment (yne) [inline src/path.js:272]: were is `flipY` declared ?

File: src/path.js (hunk 4 above)
Comment (yne) [reply src/path.js:268]: I still don't get why does we mixup in the same option/variable
- the offset
- the enabling
I would have used `{flipY:false}` `{flipY:true, flipOffsetY: 0.2}` if I get your algorithm correctly.

Comment (Connum) [reply src/path.js:268]: So we *are* talking about the option itself, not just the variable?
I think it's just simpler to have one option for this. Either you disable it, have it calculated automatically, or you provide a value. You would never set `{flipY: false, flipOffsetY: 123}`, so in my mind it didn't make sense to have this constellation.
@ILOVEPIE what do you think? In the end, it doesn't really matter to me, so I'll go with the majority.

Comment (Connum) [reply src/path.js]: split into `flipY` and `flipYBase` now, upated README accordingly

**Comment by Connum** [comment]
up-to-date with the current master and ready for re-review

**Comment by ILOVEPIE** [comment]
> I was hesitant whether to mark it as breaking or not. Having upside down output is obviously not what anyone would expect, on the other hand it's been like that for years, so there's probably code out there that relies on it and flips it itself.

This is a breaking change as it changes some public facing apis, not just wither the SVG is upside down.

**Comment by Connum** [comment]
You can still pass a number instead of the options object and it will be handled accordingly, or am I missing something else?
We also have the platform agnostic names table change in the master already and had the version field bumped to 2.0.0 before it was removed with the overhauled build/release system. So I think we should be able to merge this?
If we really need a hotfix 1.x release before 2.0.0 is ready, we can still branch it off from an earlier point before all the new features that were added recently.

**Review by yne — APPROVED** [review]

**Comment by yne** [comment]
@Connum 
I rebased-squashed your branch so it shall be mergable now,
please avoid merging master into your branch as it make the rebase more painful.

**Comment by Connum** [comment]
Can we please talk about the roadmap and how to proceed with this? I'm not that keen on resolving endless conflicts should this get merged after other PRs with big changes. 🙂 Also this SVG fix is very helpful when testing new features/enhancements/fixes with the Unicode test suite.

It was my understanding that we were going to release 2.0.0 next anyway (though I have to admit I'm not sure what that feeling is based on, as we never explicitly talked about it before) but if that's not the plan, we could at least have a dedicated 2.0.0 branch? Then I'll also scrap the partial backwards compatibility.

**Comment by yne** [comment]
@ILOVEPIE can we process?

File: src/path.js (hunk 3 above)
Comment (ILOVEPIE) [reply src/path.js]: @yne the reason why it's the same on chrome is that chrome rewrites the code to actually be ``x & 1`` when it sees ``x % 2``.

File: src/path.js [hunk 9]
```diff
@@ -16,6 +16,290 @@ function Path() {
     this.strokeWidth = 1;
 }
 
+function roundDecimal(float, places) {
+    return +(Math.round(float + 'e+' + places) + 'e-' + places);
```
Comment (ILOVEPIE) [inline src/path.js:20]: The type conversions here may slow this down. Might I suggest
```js
const shift = Math.pow(10,places);
return (Math.round(float*shift)/shift);
```
instead.

**Review by ILOVEPIE — APPROVED** [review]
There is one comment I had regarding performance/readability, but it looks good.

File: src/path.js (hunk 9 above)
Comment (Connum) [reply src/path.js]: That seems to have introduced a rounding issue

Comment (Connum) [reply src/path.js]: I'll take a closer look at this and run the unicode test suite to see what impact it has. Maybe it could even make things better in some cases. We'll see.

Comment (yne) [reply src/path.js]: From what I see, the `roundDecimal` is used at 2 places, for different reasons: 
- at path parsing, where all precision shall be kept (?)
- at serialisation, where it shall treated number as (truncated) String, not a rounded Number (which is sometime impossible due to floating point precision)

Comment (Connum) [reply src/path.js]: So when I adapt the test case, we don't fail any other tests and the number of passed tests in the Unicode test suite doesn't change, but I really think `M0.575` should change to `M0.58` when changing the `decimalPlaces` option from 3 to 3, and not `M0.57`. Why would we ever want to cut off decimal places instead of actual rounding?
I think that's why I chose to do it the way I did in the first place, because I always ended up with some weird rounding errors. But I'm open to a solution that feels less hacky and still produces correct values.

Comment (Connum) [reply src/path.js:20]: for now, I reversed the change and rebased on master

Comment (ILOVEPIE) [reply src/path.js:20]: That should be functionally equivalent to the rounding function you had originally as they do the exact same thing.

Comment (Connum) [reply src/path.js:20]: I don't know why the test is failing with a wrong value then, either 🤔

Comment (Connum) [reply src/path.js:20]: I was able to get around that rounding issue by adding `Number.EPSILON` before multiplying
```js
const shift = Math.pow(10, places);
return (Math.round((float + Number.EPSILON) * shift) / shift);
```
which makes all tests pass again - but when I tried with different values, `10.075` rounds to `10.07` instead of `10.08`.
My original implementation was the only one so far that seems to work for all cases. I'd say we leave it like that before introducing precision issues.
